Loop/Managers/Alerts/AlertStore.swift

```diff
@@ -0,0 +1,150 @@
+//
+//  AlertStore.swift
+//  Loop
+//
+//  Created by Rick Pasetto on 5/11/20.
+//  Copyright © 2020 LoopKit Authors. All rights reserved.
+//
+
+import CoreData
+import LoopKit
+
+public class AlertStore {
+    
+    public enum AlertStoreError: Error {
+        case notFound
+    }
+    
+    // Available for tests only
+    let managedObjectContext: NSManagedObjectContext
+
+    private let persistentContainer: NSPersistentContainer
+        
+    private let log = DiagnosticLog(category: "AlertStore")
+
+    private let dataAccessQueue = DispatchQueue(label: "com.loop.AlertStore.dataAccessQueue", qos: .utility)
+    
+    public init(storageFileURL: URL? = nil) {
+        managedObjectContext = NSManagedObjectContext(concurrencyType: .privateQueueConcurrencyType)
+        managedObjectContext.mergePolicy = NSMergeByPropertyStoreTrumpMergePolicy
+        managedObjectContext.automaticallyMergesChangesFromParent = true
+
+        let storeDescription = NSPersistentStoreDescription()
+        if let storageFileURL = storageFileURL {
+            storeDescription.url = storageFileURL
+        } else {
+            storeDescription.type = NSInMemoryStoreType
+        }
+        persistentContainer = NSPersistentContainer(name: "AlertStore")
+        persistentContainer.persistentStoreDescriptions = [storeDescription]
+        persistentContainer.loadPersistentStores { _, error in
+            if let error = error {
+                fatalError("Unable to load persistent stores: \(error)")
+            }
+        }
+        managedObjectContext.persistentStoreCoordinator = persistentContainer.persistentStoreCoordinator
+    }
+}
+
+// MARK: Alert Recording
+
+extension AlertStore {
+    
+    public func recordIssued(alert: DeviceAlert, at date: Date = Date(), completion: ((Result<Void, Error>) -> Void)? = nil) {
```

Comment:
I saw Darin pointed out that this is taking `DeviceAlert`.  It could accept a separate protocol or struct called `NewAlert` (we have a similar pattern with NewPumpEvent).  If it's a protocol, then DeviceAlert could just conform to it. The "New" in the name is not perfect, as we use the same structs in the pump apis to update existing records, and I'd think you'd do the same here. But it's just an idea.

Reply:
Honestly, I had thought of this when I first made `DeviceAlert`.  You may recall that my initial name for that struct was `UserAlert` (meaning: an alert for the _User_).  There really is nothing "device-ish" about it...the only thing coming close is the `managerIdentifier` (which, conveniently and somewhat intentionally, also doesn't have the word _device_ in it). 

I have nothing against protocols (love them! I sprinkle them on my breakfast!), but in this case, I'm not sure I see the value.  They are good when abstracting complex objects (managers, delegates, etc.) but in this case it is really just a package of immutable information, an ideal choice for a `struct` IMHO. (In fact, `DeviceAlert` has no external dependencies...it's just nested structs of primitive types).

I think I'd much rather change `DeviceAlert` to just `Alert`, and then these names will all match.  The downside of course is that this name matches a `SwiftUI` object as well as (I think) an object in the DexcomSDK.  However, the Swift compiler should protect us in that case (e.g. `SwiftUI.Alert` vs `LoopKit.Alert`). 

WDYT?  If you agree, I'll make a follow-up PR

Reply:
I think that's fine; if there are no special distinguishing factors between alerts issued by devices and general system alerts. Only thing I can think of is tracking where it came from.  Your central alert manager needs to send acknowledgements for alerts from devices back to their device managers, and you don't need to do that for system alerts. If fulfilling that responsibility entails changing the alert structure (adding a device manager id, for example), then there is some difference.

Reply:
I'm guessing your plan is just to keep that device manager id in all the contexts (app level, storage, and devicemanager interface), and I think that's ok. I tend to think having specific structs for specific contexts is nice because it helps keep concerns from spreading, and provides helps keeps the boundaries explicit. But the cost is extra structs, and conversions between them. Sometimes that's nice in terms of organizing and keeping each context clean, and sometimes it's just extra work.

Reply:
Yes, my plan is to keep `managerIdentifier`, not necessarily meaning "device manager".  I hadn't thought of `managerIdentifier` as anything other than a namespace scoping (i.e. "this alert came from that manager") -- but that might be because of my only recently working in the codebase. 
When we start doing "system" alerts, I imagine they'll need to be scoped as well.  One might come from the "loop algorithm" area (which may be in LoopKit), another from the "system level" (e.g. disk space low) area (which may be in Loop), and yet others from (backend) Services (which are separate plugins, though I can't come up with what alert we would want from them). 
The `Identifier` structure is meant to be exactly the rendezvous mechanism for alert acknowledgement, so we know where to send what acknowledgment.
Perhaps the name `managerIdentifier` is the place we're getting tripped up?
At any rate, I understand you concern, and I can see how to go the other route should we choose it.  
(I'll re-request your approval for all of this.  If you still disagree, happy to chat more about it.)

Reply:
The AlertManager at the LoopLevel is within the context of Loop itself, and it would make complete sense for it to be the "handler" for an acknowledgment, and would not need further routing.  But again, this is a small thing, and I think it's acceptable having a global alert with a device identifier in it even for alerts that don't need further routing.